core/Tuples.carp

```diff
@@ -1,43 +1,95 @@
-  (defn zero []
-      (Pair.init (zero) (zero)))
+(defmodule Dynamic
+  (private deftuple-type-)
+  (hidden deftuple-type-)
+  (defndynamic deftuple-type- [name props]
+    (list 'deftype (cons name props)
+      (collect-into (flatten (map (fn [x] (list x x)) props)) array)))
+
+  (private deftuple-lt-)
+  (hidden deftuple-lt-)
+  (defndynamic deftuple-lt- [name props]
+    (if (empty? props)
+      'false
+      (let [fst (Symbol.prefix name (car props))]
+        (if (= (length props) 1)
+          (list '< (list fst 't1) (list fst 't2))
+          (list 'if (list '= (list fst 't1) (list fst 't2))
+            (deftuple-lt- name (cdr props))
+            (list '< (list fst 't1) (list fst 't2)))))))
+
+  ; this is basically just a giant template
+  (private deftuple-module-)
+  (hidden deftuple-module-)
+  (defndynamic deftuple-module- [name props]
+    (let [sname (Symbol.str name)
+          module-name (Symbol.concat [name 'Ref])]
+      (list 'do
+        (list 'defmodule module-name
+          (list 'defn '= ['t1 't2]
+            (cons 'and*
+              (map (fn [p]
+                    (list '=
+                      (list (Symbol.prefix name p) 't1)
+                      (list (Symbol.prefix name p) 't2)))
+                   props)))
+          (list 'implements '= (Symbol.prefix module-name '=))
+
+          (list 'defn '< ['t1 't2] (deftuple-lt- name props))
+          (list 'implements '< (Symbol.prefix module-name '<))
+
+          (list 'defn '> ['t1 't2] (list (Symbol.prefix module-name '<) 't2 't1))
+          (list 'implements '> (Symbol.prefix module-name '>)))
+
+        (list 'defmodule name
+          (list 'doc 'init-from-refs
+            (String.concat ["initializes a `" sname "` from member references."]))
+          (let [prop-vars (map (fn [x] (Symbol.concat [x '-val])) props)]
+            (list 'defn 'init-from-refs (collect-into prop-vars array)
+              (cons 'init (map (fn [x] (list 'copy x)) prop-vars))))
+
+          (list 'defn '= ['t1 't2]
+            (cons 'and*
+              (map (fn [p] (list '= (list p '(ref t1)) (list p '(ref t2)))) props)))
+          (list 'implements '= (Symbol.prefix name '=))
+
+          (list 'defn '< ['t1 't2]
+            (list (Symbol.prefix module-name '<) '(ref t1) '(ref t2)))
+          (list 'implements '< (Symbol.prefix name '<))
+
+          (list 'defn '> ['t1 't2]
+            (list (Symbol.prefix module-name '>) '(ref t1) '(ref t2)))
+          (list 'implements '> (Symbol.prefix name '>))
+
+          (list 'doc 'reverse
+            (String.concat ["reverses a `" sname "` by reversing its member positions."]))
+          (list 'defn 'reverse ['t]
+            (cons 'init (map (fn [x] (list 'copy (list x 't))) (reverse props))))
+
+          (list 'defn 'zero [] (cons 'init (map (fn [_] '(zero)) props))))
+          (list 'implements 'zero (Symbol.prefix name 'zero))
+          ;(list 'doc 'zero
```

Comment:
For some reason, removing the comment here makes the compiler loop. The `doc` commands above don’t do that, so I’m not quite sure what’s happening there. Is it because it’s an `interface`?

Reply:
I suspect it has to do with the logic in `MetaSet`--the lookup is a bit messy atm

```haskell
let env = contextGlobalEnv ctx
      pathStrings = contextPath ctx
      fppl = projectFilePathPrintLength (contextProj ctx)
      fullPath = consPath pathStrings path
  case lookupInEnv fullPath env of
    Just (foundEnv, binder@(Binder _ xobj)) ->
      -- | Set meta on existing binder
      setMetaOn ctx (Just foundEnv) binder
```

Normally, consing the path should make it resolve to the right module-specific impl of `zero`, but there are issues with submodules and contextstack popping whereby for some reason MetaSet only gets the end of the stack. My guess is that the issue here might be similar. Either that or probably the local `setMetaOn` causes an infinite loop via it's different approach to resolving the pathname:

```
 setMetaOn :: Context -> Maybe Env -> Binder -> IO (Context, Either EvalError XObj)
      setMetaOn ctx foundEnv binder@(Binder metaData xobj) =
        do let globalEnv = contextGlobalEnv ctx
               newMetaData = MetaData (Map.insert key value (getMeta metaData))
               xobjPath = getPath xobj
               prefixPath = fromMaybe [] (fmap pathToEnv foundEnv)
               fullPath = case xobjPath of
                            SymPath [] _ -> consPath prefixPath xobjPath
                            SymPath _ _ -> xobjPath
```

Probably we should only resolve the path in one place, and not both the main and local `setMetaOn` bindings.

Just a hunch, not sure if this is really the root cause yet. It's also possible simply looking in the Global Env causes it to find the interface first and there's some problems with setting docs on that (shouldn't happen if we reconstruct the proper fullpath, but again, there are submodule related issues to getting the right fullpath in MetaSet, so I wouldn't be surprised if the same problems sometimes occur in edge cases for non-submodules.)

Presumably, we could resolve this doing `doc module-name.zero` -- but forward declarations don't support qualified names yet.

Reply:
Yeah your hunch seems spot on, maybe it's something we can fix by doing the meta data cleanup we've discussed earlier..?

Reply:
Yep, sounds good! I'll hopefully get around to Meta cleanup sometime this week.